File: tempIdeas/java/src/Cristallo.java
import java.util.List;

public class Cristallo extends Pietra {
    private String magia_;



    //CONSTRUCTORS
    public Cristallo() {
        this(1,1,1.0,1.0,1.0);
    }

    public Cristallo(Integer livello, Integer rarita, Double spirito, Double durezza, Double magia) {
        super(livello, rarita, spirito, durezza);
        magia_="Magia";
        insertStat(magia_,magia);

    }

    //METODI DI GET
    Double getMagia() {
        return getValoreStat(magia_);
    }

    public Cristallo clone() {
        return new Cristallo(getLivello(), getRarita(), getSpirito(), getDurezza(), getMagia());
    }

    @Override
    public void potenzia(Double mana) {
        mana=sanitizeMana(mana);
        Double incremento = mana*getLivello()*getRarita();
        Integer divisore=10;
        incrementStat(magia_, incremento*getRarita()/divisore);
        super.potenzia(mana);
        normalizza();
    }


    @Override
    public void potenzia(Double mana, String parametro) {
        if(getListaStats().contains(parametro)) {
            mana = sanitizeMana(mana);
            Double incremento = mana * getLivello() * getRarita();
            Integer divisore = 1;

            if (parametro == "Spirito") {
                divisore = 15;
                incrementStat(parametro, incremento / getMagia());
                incrementStat(magia_, incremento * getRarita() / divisore);
            } else {
                divisore = 10;
                incrementStat(magia_, incremento * getRarita() / divisore);
                super.potenzia(mana, parametro);
            }
            normalizza();
        } else {
            this.potenzia(mana);
        }
    }

    public void estraiDa(Oggetto oggetto) throws OperationException {
        try {
            if(oggetto instanceof Cristallo) throw new OperationException(OperationException.ecc.estrazione);
            Pietra p = (Pietra) oggetto;

            if (p.getDurezza() >= 10) {
                setLivello(p.getLivello());
                setRarita(p.getRarita());

                List<String> s=getListaStats();
                Integer numeroStats = s.size();
                final Double sommaStats = p.getSommaStats();

                s.stream().forEach(f -> {
                    if (f == magia_) incrementStat(f, (sommaStats - p.getSpirito()) / numeroStats);
                    else incrementStat(f, p.getValoreStat(f) * (numeroStats - 1) / numeroStats);
                });
            } else
                getListaStats().forEach(s -> modifyStat(s, 1.0));
        } catch (ClassCastException e) {
            throw new OperationException(OperationException.ecc.estrazione);
        }
    }

    @Override
    public Double ricicla() {
        return super.ricicla() + getMagia()*getRarita();
    }

    public void distribuisci(Oggetto obj) {
        Double val=getMagia()/2;
        if(getDurezza() < val +1) return;

        editDurezza(-val);

        List<String> lista=obj.getListaStats();
        Integer size=lista.size();

        lista
                .stream()
                .forEach(s->
                        obj.incrementStat(s, val/size)
                );
        obj.normalizza();
    }
}

File: tempIdeas/untitled/cristallo.cpp
#include "cristallo.h"

Cristallo::Cristallo(const int livello, const int rarita, const double spirito, const double durezza, const double magia) : Pietra(livello, rarita, spirito, durezza), magia_("Magia") {
    Oggetto::insertStat(magia_, magia);
}

double Cristallo::getMagia() const {
    return getValoreStat(magia_);
}

Cristallo *Cristallo::clone() const {
    return new Cristallo(*this);
}

void Cristallo::potenzia(double mana, std::string parametro) {

    mana = sanitizeMana(mana);

    double incremento = mana * getLivello() * getRarita();
    int divisore =1;

    if(parametro == "Spirito") {
        divisore = 15;
        incrementStat(parametro, incremento/getMagia());
        incrementStat(magia_, incremento * getRarita() / divisore);
    }
    else {
        divisore = 10;
        incrementStat(magia_, incremento * getRarita() / divisore);
        Pietra::potenzia(mana, parametro);
    }
    normalizza();
}

double Cristallo::ricicla() const {
    return Pietra::ricicla() + getMagia() * getRarita();
}

void Cristallo::estraiDa(Oggetto *oggetto) {
    if(typeid(Pietra) == typeid(*oggetto)) {

        list<string> s = getListaStats();

        if((static_cast<Pietra*>(oggetto))->getDurezza() >= 10) {
            setLivello(oggetto->getLivello());
            setRarita(oggetto->getRarita());

            int numeroStat = s.size();

            for(auto i = s.begin(); i != s.end(); ++i)
                if(*i == magia_) incrementStat(*i, (oggetto->getSommaStats() - oggetto->getSpirito()) / numeroStat);
                else incrementStat(*i, oggetto->getValoreStat(*i) * (numeroStat - 1) / numeroStat );
        }
        else {
            for(auto i = s.begin(); i != s.end(); ++i)
                modifyStat(*i, 1);
        }
    }
    else {
        throw OperationException(OperationException::estrazione);
    }
}

void Cristallo::distribuisci(Oggetto *obj) {
    double val = getMagia() / 2;

    if(getDurezza() < val + 1) {
        return;
    }

    editDurezza(-val);

    list<string> parametri = obj->getListaStats();

    for(auto it = parametri.begin(); it != parametri.end(); ++it)
        obj->incrementStat(*it, val / parametri.size());

    obj->normalizza();
}
